// office-ui-fabric-react/lib-amd/components/SwatchColorPicker/SwatchColorPicker.js
define(["require", "exports", "tslib", "react", "../../Utilities", "../../utilities/color/colors", "../../utilities/grid/Grid", "../../Button", "../../FocusZone", "./SwatchColorPicker.scss"], function (require, exports, tslib_1, React, Utilities_1, colors_1, Grid_1, Button_1, FocusZone_1, stylesImport) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var styles = stylesImport;
    var SwatchColorPickerOption = (function (_super) {
        tslib_1.__extends(SwatchColorPickerOption, _super);
        function SwatchColorPickerOption(props) {
            var _this = _super.call(this, props) || this;
            _this._id = props.id || Utilities_1.getId('colorCell');
            return _this;
        }
        SwatchColorPickerOption.prototype.render = function () {
            var _a = this.props, item = _a.item, id = _a.id, className = _a.className, role = _a.role, selectedIndex = _a.selectedIndex, disabled = _a.disabled, cellShape = _a.cellShape, onClick = _a.onClick, onHover = _a.onHover, onFocus = _a.onFocus;
            return (React.createElement(Button_1.CommandButton, { id: id + '-item' + item.index, "data-index": item.index, "data-is-focusable": true, disabled: disabled, className: Utilities_1.css(className, (_b = {},
                    _b['is-selected ' + styles.cellIsSelected] = (selectedIndex !== undefined && selectedIndex === item.index),
                    _b['is-disabled ' + styles.disabled] = disabled,
                    _b)), onClick: this._onClick, onMouseEnter: this._onMouseEnter, onMouseLeave: this._onMouseLeave, onFocus: this._onFocus, role: role, "aria-selected": selectedIndex !== undefined && (selectedIndex === item.index).toString(), ariaLabel: item.label && item.label, title: item.label && item.label }, this._onRenderOption()));
            var _b;
        };
        SwatchColorPickerOption.prototype._onClick = function () {
            var _a = this.props, onClick = _a.onClick, disabled = _a.disabled, item = _a.item;
            if (onClick && !disabled) {
                onClick(item);
            }
        };
        SwatchColorPickerOption.prototype._onMouseEnter = function () {
            var _a = this.props, onHover = _a.onHover, disabled = _a.disabled, item = _a.item;
            if (onHover && !disabled) {
                onHover(item.id, item.color);
            }
        };
        SwatchColorPickerOption.prototype._onMouseLeave = function () {
            var _a = this.props, onHover = _a.onHover, disabled = _a.disabled;
            if (onHover && !disabled) {
                onHover();
            }
        };
        SwatchColorPickerOption.prototype._onFocus = function () {
            var _a = this.props, onFocus = _a.onFocus, disabled = _a.disabled, item = _a.item;
            if (onFocus && !disabled) {
                onFocus(item.id, item.color);
            }
        };
        /**
         * Render the core of a color cell
         * @returns {JSX.Element} - Element representing the core of the item
         */
        SwatchColorPickerOption.prototype._onRenderOption = function () {
            var _a = this.props, cellShape = _a.cellShape, item = _a.item;
            // Build an SVG for the cell with the given shape and color properties
            return (React.createElement("svg", { className: Utilities_1.css(styles.svg, cellShape, cellShape === 'circle' ? styles.circle : ''), viewBox: '0 0 20 20', fill: colors_1.getColorFromString(item.color).str }, cellShape === 'circle' ?
                React.createElement("circle", { cx: '50%', cy: '50%', r: '50%' }) :
                React.createElement("rect", { width: '100%', height: '100%' })));
        };
        SwatchColorPickerOption.defaultProps = {
            cellShape: 'circle',
            disabled: false
        };
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPickerOption.prototype, "_onClick", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPickerOption.prototype, "_onMouseEnter", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPickerOption.prototype, "_onMouseLeave", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPickerOption.prototype, "_onFocus", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPickerOption.prototype, "_onRenderOption", null);
        return SwatchColorPickerOption;
    }(React.Component));
    var SwatchColorPicker = (function (_super) {
        tslib_1.__extends(SwatchColorPicker, _super);
        function SwatchColorPicker(props) {
            var _this = _super.call(this, props) || this;
            _this._id = props.id || Utilities_1.getId('swatchColorPicker');
            var selectedIndex;
            if (props.selectedId) {
                selectedIndex = _this._getSelectedIndex(props.colorCells, props.selectedId);
            }
            _this.state = {
                selectedIndex: selectedIndex
            };
            return _this;
        }
        SwatchColorPicker.prototype.componentWillReceiveProps = function (newProps) {
            var newSelectedIndex;
            if (newProps.selectedId) {
                newSelectedIndex = this._getSelectedIndex(newProps.colorCells, newProps.selectedId);
            }
            if (newSelectedIndex !== undefined &&
                newSelectedIndex !== this.state.selectedIndex) {
                this.setState({
                    selectedIndex: newSelectedIndex
                });
            }
        };
        SwatchColorPicker.prototype.render = function () {
            var _a = this.props, disabled = _a.disabled, colorCells = _a.colorCells, columnCount = _a.columnCount, positionInSet = _a.positionInSet, setSize = _a.setSize, shouldFocusCircularNavigate = _a.shouldFocusCircularNavigate, className = _a.className, onCellFocused = _a.onCellFocused;
            return (React.createElement(FocusZone_1.FocusZone, { isCircularNavigation: shouldFocusCircularNavigate, className: Utilities_1.css('ms-swatchColorPickerBodyContainer', styles.swatchColorPickerContainer, className), onBlur: this.onSwatchColorPickerBlur },
                React.createElement(Grid_1.Grid, { key: this._id + colorCells[0].id + '-grid', items: colorCells.map(function (item, index) { return tslib_1.__assign({}, item, { index: index }); }), columnCount: this.props.columnCount, onRenderItem: this._renderOption, positionInSet: positionInSet && positionInSet, setSize: setSize && setSize })));
        };
        /**
         * When the whole swatchColorPicker is blurred,
         * make sure to clear the pending focused stated
         */
        SwatchColorPicker.prototype.onSwatchColorPickerBlur = function () {
            if (this.props.onCellFocused) {
                this.props.onCellFocused();
            }
        };
        /**
         * Get the selected item's index
         * @param items - The items to search
         * @param selectedId - The selected item's id to find
         * @returns {number} - The index of the selected item's id, -1 if there was no match
         */
        SwatchColorPicker.prototype._getSelectedIndex = function (items, selectedId) {
            var selectedIndex = Utilities_1.findIndex(items, (function (item) { return (item.id === selectedId); }));
            return selectedIndex >= 0 ? selectedIndex : undefined;
        };
        /**
         * Render a color cell
         * @param item - The item to render
         * @returns {JSX.Element} - Element representing the item
         */
        SwatchColorPicker.prototype._renderOption = function (item) {
            var id = this._id;
            return (React.createElement(SwatchColorPickerOption, { item: item, id: id, key: id + item.id, disabled: this.props.disabled, cellShape: this.props.cellShape, className: styles.cell, onClick: this._onCellClick, onHover: this.props.onCellHovered, onFocus: this.props.onCellFocused, role: 'gridcell', selectedIndex: this.state.selectedIndex }));
        };
        /**
         * Handle the click on a cell
         * @param item - The cell that the click was fired against
         */
        SwatchColorPicker.prototype._onCellClick = function (item) {
            if (this.props.disabled) {
                return;
            }
            var index = item.index;
            // If we have a valid index and it is not already
            // selected, select it
            if (index >= 0 && index !== this.state.selectedIndex) {
                if (this.props.onColorChanged) {
                    this.props.onColorChanged(item.id, item.color);
                }
                this.setState({
                    selectedIndex: index
                });
            }
        };
        /**
         * Clear the colors by calling the given callbacks
         * @param callbacks - The callbacks to handle the clear operation
         */
        SwatchColorPicker.prototype._clearColors = function (callbacks) {
            callbacks.forEach(function (callback) {
                if (callback) {
                    callback();
                }
            });
        };
        SwatchColorPicker.defaultProps = {
            cellShape: 'circle',
            disabled: false,
            shouldFocusCircularNavigate: true
        };
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPicker.prototype, "onSwatchColorPickerBlur", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPicker.prototype, "_renderOption", null);
        tslib_1.__decorate([
            Utilities_1.autobind
        ], SwatchColorPicker.prototype, "_onCellClick", null);
        return SwatchColorPicker;
    }(Utilities_1.BaseComponent));
    exports.SwatchColorPicker = SwatchColorPicker;
});
