// office-ui-fabric-react/lib-amd/components/CommandBar/examples/CommandBar.Example.scss.js
define(["require", "exports", "@microsoft/load-themed-styles"], function (require, exports, load_themed_styles_1) {
    "use strict";
    var styles = {
        customButtonContainer: 'customButtonContainer_1741ce49',
        icon: 'icon_1741ce49',
        themeDarkAltColor: 'themeDarkAltColor_1741ce49',
        leftText: 'leftText_1741ce49',
        splitter: 'splitter_1741ce49',
        button: 'button_1741ce49',
        darkerBG: 'darkerBG_1741ce49',
    };
    load_themed_styles_1.loadStyles([{ "rawString": ".customButtonContainer_1741ce49{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center}.icon_1741ce49{display:inline-block;font-family:FabricMDL2Icons;font-style:normal;font-weight:400;speak:none;font-size:14px;position:relative}.themeDarkAltColor_1741ce49{color:" }, { "theme": "themeDarkAlt", "defaultValue": "#106ebe" }, { "rawString": "}.leftText_1741ce49{margin-left:6px;font-size:14px;font-weight:400;color:" }, { "theme": "neutralPrimary", "defaultValue": "#333333" }, { "rawString": "}.splitter_1741ce49{font-size:14px;font-weight:400;color:" }, { "theme": "neutralTertiary", "defaultValue": "#a6a6a6" }, { "rawString": ";background:0 0;position:relative;top:-2px;height:40px}.button_1741ce49.ms-Button{height:40px;min-width:28px;padding:0 5px}.button_1741ce49:hover{background:initial}.darkerBG_1741ce49{background:#d0d0d0}" }]);
    return styles;
});
/* tslint:enable */
